### PR TITLE
first proposal of cv::remap with relative displacement field (#24603)

### PR DESCRIPTION
**Pull request by chacha21**
Implements #24603

Currently, `remap()` is applied as `dst(x, y) <- src(mapX(x, y), mapY(x, y))` It means that the maps must be filled with absolute coordinates.

However, if one wants to remap something according to a displacement field ("warp"), the operation should be `dst(x, y) <- src(x+displacementX(x, y), y+displacementY(x, y))`

It is trivial to build a mapping from a displacement field, but it is an undesirable overhead for CPU and memory.

This PR implements the feature as an experimental option, through the optional flag WARP_RELATIVE_MAP than can be ORed to the interpolation mode.

Since the xy maps might be const, there is no attempt to add the coordinate offset to those maps, and everything is postponed on-the-fly to the very last coordinate computation before fetching `src`. Interestingly, this let `cv::convertMaps()` unchanged since the fractional part of interpolation does not care of the integer coordinate offset.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [X] I agree to contribute to the project under Apache 2 License.
- [X] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [X] The PR is proposed to the proper branch
- [X] There is a reference to the original bug report and related work
- [X] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by asmorkalov** [comment]
@vpisarev Friendly reminder.

**Comment by asmorkalov** [comment]
@vpisarev Friendly reminder.

**Review by asmorkalov — CHANGES_REQUESTED** [review]
Thanks a lot for the great job! The proposed option is really very useful. I apologize for large delay caused by release and some 5.0 preparation activities. General proposals:
-  Extend accuracy tests. The current test does not cover all touched cases. Also the test scenario looks very simple. As soon as absolute->relative displacement conversion is very simple the test expansion may be done easily.
- Need to add several  corner cases, e.g. absolute coordinate is close to type range
- There are no performance tests. I propose to patch the existing one as in the first item.

File: modules/imgproc/src/imgwarp.cpp [hunk 1]
```diff
@@ -326,11 +326,12 @@ static inline int clip(int x, int a, int b)
...
 \****************************************************************************************/
 
-template<typename T>
+template<typename T, bool isRelative>
 static void remapNearest( const Mat& _src, Mat& _dst, const Mat& _xy,
-                          int borderType, const Scalar& _borderValue )
+                          int borderType, const Scalar& _borderValue, const Point& _offset )
 {
     Size ssize = _src.size(), dsize = _dst.size();
+    const Point offset = _offset;
```
[the remark this answers is not in the document]
Comment (chacha21) [reply modules/imgproc/src/imgwarp.cpp:334]: What about performance ? don't you think that keeping a reference sill prevent the compiler to optimize access to offset.x|y in the inner loops ?

File: modules/imgproc/src/imgwarp.cpp [hunk 2]
```diff
@@ -1385,9 +1426,9 @@ static bool ocl_remap(InputArray _src, OutputArray _dst, InputArray _map1, Input
             scalararg = ocl::KernelArg::Constant((void*)scalar.ptr(), scalar.elemSize());
 
     if (map2.empty())
-        k.args(srcarg, dstarg, map1arg, scalararg);
+        k.args(srcarg, dstarg, map1arg, scalararg, (char)hasRelativeFlag);
     else
-        k.args(srcarg, dstarg, map1arg, ocl::KernelArg::ReadOnlyNoSize(map2), scalararg);
+        k.args(srcarg, dstarg, map1arg, ocl::KernelArg::ReadOnlyNoSize(map2), scalararg, (char)hasRelativeFlag);
```
[the remark this answers is not in the document]
Comment (chacha21) [reply modules/imgproc/src/imgwarp.cpp:1431]: Ok, I recently learnt how to to that

File: modules/imgproc/src/imgwarp.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (asmorkalov) [reply modules/imgproc/src/imgwarp.cpp:334]: good question. we need performance test for it ;)

**Comment by chacha21** [comment]
A few comments after the last commit :

- do we keep the "WARP_RELATIVE_MAP" flag ? Is there a better strategy to enable that code ? Is the name OK ?
- using the corner case "absolute coordinate is close to type range" : using some `saturate_add` would be great, but AFAIK, only OpenCL provides such an operator so far
- about the "local variable", there usage will be determined by perf tests, but...
- ...I don't know how to write and run perf tests. I have never been able to to that on my development machine with Windows7 and broken Python.

**Comment by asmorkalov** [comment]
do we keep the "WARP_RELATIVE_MAP" flag ? Is there a better strategy to enable that code ? Is the name OK ? (I have not mentioned it in the doc yet) - Looks good to me.

**Comment by vpisarev** [comment]
@chacha21, thank you for the contribution!

I like how OpenCL part is implemented. It's conditionally compiled code, and so it does not affect performance of the standard case. But I don't like that in CPU version there are extra conditions inside the innermost loops. And the extra registers needed to hold the pixel grid coordinates. In subsequent versions of OpenCV we would like to optimize remap further, not to slow it down. We want to keep it clean, we want to avoid any unnecessary overhead.

Here is the proposed solution. If "relative" flag is set, a tile of map(s) should be copied to a temporary buffer (probably stack-allocated buffer) and augmented there prior to calling the remap kernels instead of doing it in the remap kernels themselves. Since the offsets for x and y are integers, such method is compatible with both floating-point and the fixed-point representations of the maps.

**Comment by opencv-alalek** [comment]
> But I don't like that in CPU version there are extra conditions inside the innermost loops

There is no such problem because there is `template<bool isRelative>` in the most critical paths.

---

> to optimize remap further, not to slow it down

Just need to provide performance report for such modifications.
All PRs which provides optimization or modification of implementation should have that report.

**Comment by chacha21** [comment]
> > But I don't like that in CPU version there are extra conditions inside the innermost loops

> There is no such problem because there is `template<bool isRelative>` in the most critical paths.

Exactly, my first commit on this PR was a local `bool isRelative` and I switched to template version after to keep performance.

> > to optimize remap further, not to slow it down
> 
> Just need to provide performance report for such modifications. All PRs which provides optimization or modification of implementation should have that report.

I will try to run the test and report ASAP (not familiar at all with the procedure)

**Comment by vpisarev** [comment]
@chacha21, yes, I'm probably wrong about performance - I still see some unconditional things like vector registers holding the pixel coordinates. Maybe compiler will optimize it out, but maybe not. Besides, it basically duplicates all the remap kernels for such a very rarely used feature. I'd still suggest to do it externally by copying each tile of maps into a temporary buffer and augmenting it there. The kernels then will stay unchanged.

**Review by vpisarev — CHANGES_REQUESTED** [review]
please, do mapx & mapy augmentation as a separate preprocessing step (probably, tile-by-tile, to achieve cache and thread locality), not inside interpolation kernels

**Comment by chacha21** [comment]
@vpisarev
I understand, but I am not sure that tiling maps to create "augmented" copies will be better.
As far as I understand, the best place would be in an alternative `RemapInvoker` (perhaps a `template<bool isRelative> RemapInvoker`) to get rid of the `template `of the remap kernels.
So the tiling would occur in the `virtual void RemapInvoker::operator()`, something like that :

Actual :
```
        int x, y, x1, y1;
        const int buf_size = 1 << 14;
        int brows0 = std::min(128, dst->rows), map_depth = m1->depth();
        int bcols0 = std::min(buf_size/brows0, dst->cols);
        brows0 = std::min(buf_size/bcols0, dst->rows);

        Mat _bufxy(brows0, bcols0, CV_16SC2), _bufa;
        if( !nnfunc )
            _bufa.create(brows0, bcols0, CV_16UC1);
```
modified : 

```
        int x, y, x1, y1;
        const int buf_size = 1 << 14;
        int brows0 = std::min(128, dst->rows), map_depth = m1->depth();
        int bcols0 = std::min(buf_size/brows0, dst->cols);
        brows0 = std::min(buf_size/bcols0, dst->rows);

        Mat m1AugmentedTile(brows0 , bcols0, m1->type());
        Mat m2AugmentedTile(brows0 , bcols0, m1->type());
        fillByAddingRelativeOffset(m1AugmentedTile, m1);
        fillByAddingRelativeOffset(m2AugmentedTile, m2);
        //then in the code below, use m1AugmentedTile and m2AugmentedTile instead of m1, m2

        Mat _bufxy(brows0, bcols0, CV_16SC2), _bufa;
        if( !nnfunc )
            _bufa.create(brows0, bcols0, CV_16UC1);
```

As the first step, I just tried to observe the overhead of the allocation of m1AugmentedTile and m2AugmentedTile, with their content copied from m1 and m2 without modification.
`brows0 x bcols0`  seems a little large for an AutoBuffer, so I relied on a `Mat`.

And the timing is not good.

original implementation :
1000 x cv::remap((1280x1024)) => 1282.797557ms
1000 x cv::remap((1280x1024)+WARP_RELATIVE_MAP => 1490.316975ms (~+15% you're right it is not negligible)

When allocating m1AugmentedTile/m2AugmentedTile in `RemapInvoker::operator()` :
1000 x cv::remap((1280x1024)) => 1934.893129ms

The overhead of using m1AugmentedTile/m2AugmentedTile is from the beginning larger than the current proposal.

So I have to admit than WARP_RELATIVE_MAP  is not free. But my idea was that the cost of using relative offsets on the fly was still cheaper than creating absolute maps from relative maps before calling `remap()`. I think it still holds.

What do you think ? Shoudl I try with a stack allocation for m1AugmentedTile/m2AugmentedTile or is it a bad idea ?

**Review by asmorkalov — APPROVED** [review]
:+1:

**Review by vpisarev — APPROVED** [review]

**Comment by vpisarev** [comment]
ok, since @asmorkalov could not reproduce any regressions on his machines, let's merge it in!

**Pull request merged**